Closes #12564. Can't solve limit(x**2 + x*sin(x) + cos(x), x, -oo).

maurogaravello opened the pull request:
In some situation, for example if `expr = x**2 + x*sin(x)`, `limit(expr, x, oo)` fails
since `heuristics()` in `limits.py` calculates the limit of `x**2` and `x*sin(x)` separately.
The second one returns all the real line as limit points.
If such happens, then the procedure stops.
Instead, if `expr = x * (x + sin(x))`, then sympy gives the correct answer.
Therefore some lines to `limits.py` are added in order to check whether
`factor_terms(expr)` or `factor(expr)` can help in solving the limit.

`factor(expr)` is more expensive than `factor_terms(expr)`. Hence we inserted some conditional
in order to prevent, if possible, the use of `factor(expr)`.

Moreover, two tests are  added.

This PR is a duplicate of PR #12607. This in order to make more clear the changes done in such PR, as suggested by a comment. 

Closes #12607 

<!-- Please give this pull request a descriptive title. Pull requests with descriptive titles are more likely to receive reviews. Describe what you changed! A title that only references an issue number is not descriptive. -->

<!-- If this pull request fixes an issue please indicate which issue by typing "Fixes #NNNN" below. -->

maurogaravello committed:
This PR fix issue #12564.

In some situation, for example if expr = x**2 + x*sin(x), limit(expr, x, oo) fails
since heuristics() in limits.py calculates the limit of x**2 and x*sin(x) separately.
The second one returns all the real line as limit points.
If such happens, then the procedure stops.
Instead, if expr = x * (x + sin(x)), then sympy gives the correct answer.
Therefore some lines to limits.py are added in order to check whether
factor_terms(expr) or factor(expr) can help in solving the limit.

factor(expr) is more expensive than factor_terms(expr). Hence we inserted some conditional
in order to prevent, if possible, the use of factor(expr).

Moreover, two tests are added.

smichr commented:
I closed the other issue. Looking at what you were wanting the expr to become via factor, I can tell you that the function you should try after factor terms, before factor, is `together`.

maurogaravello commented:
What you mean by `is together`?

smichr commented:
Btw, to make this close the issue, you have to have the issue number next to a keyword. See [here](https://help.github.com/articles/closing-issues-using-keywords/).

maurogaravello commented:
So, you propose a code like this?

 ```
             if isinstance(e, Add):
                    m = factor_terms(e)
                    if not isinstance(m, Mul): 
                        m = together(m)
                    if not isinstance(m, Mul):
                        m = factor(e)
                    if isinstance(m, Mul):
                        return heuristics(m, z, z0, dir)
                    return
```
I do not see which cases works with `together`.

smichr commented:
> So, you propose a code like this?

yes. And I like this unnested form.

smichr commented:
And I suppose a test like this could be added:

```
assert limit(((x + cos(x))**2).expand(), x, oo) == oo
```

maurogaravello committed:
In some situations, for example if expr = x**2 + x*sin(x), limit(expr, x, oo) fails
since heuristics() in limits.py calculates the limit of x**2 and x*sin(x) separately.
The second one returns all the real line as limit points.
If such happens, then the procedure stops.

In this commit, following a suggestion, before trying to use factor(), sympy tries to use the
less expensive function together(). If both factor_terms() and together() fail,
then sympy passes to factor().

Moreover, four tests of the type
'assert limit(((x + cos(x))**2).expand(), x, oo) == oo' are added.

maurogaravello commented:
@smichr I modified the code according to your suggestions. I noticed that `together()` is sufficient for all the examples. More precisely, the program does not enter in the line with `factor()`. A possibility is to completely eliminate the piece of code with `factor()`. However it is not clear to me if `factor()` could be important for some particular cases.
